File: kirun-js/src/engine/function/system/object/ObjectDeleteKey.ts
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';

import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { isNullValue } from '../../../util/NullCheck';
import { duplicate } from '../../../util/duplicate';
import { AbstractFunction } from '../../AbstractFunction';

const VALUE = 'value';
const SOURCE = 'source';
const KEY = 'key';

export class ObjectDeleteKey extends AbstractFunction {
    public constructor() {
        super();
        this.signature = new FunctionSignature('ObjectDeleteKey')
            .setNamespace(Namespaces.SYSTEM_OBJECT)
            .setParameters(
                new Map([
                    Parameter.ofEntry(SOURCE, Schema.ofAny(SOURCE)),
                    Parameter.ofEntry(KEY, Schema.ofString(KEY)),
                ]),
            )
            .setEvents(
                new Map([Event.outputEventMapEntry(new Map([[VALUE, Schema.ofAny(VALUE)]]))]),
            );
    }

    private readonly signature;
    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let source = context.getArguments()?.get(SOURCE);
        let key = context.getArguments()?.get(KEY);

        if (isNullValue(source))
            return new FunctionOutput([EventResult.outputOf(new Map([[VALUE, undefined]]))]);

        source = duplicate(source);
        delete source[key];

        return new FunctionOutput([EventResult.outputOf(new Map([[VALUE, source]]))]);
    }
}

File: kirun-py/src/kirun_py/function/system/object/object_delete_key.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces
from kirun_py.util.duplicate import duplicate
from kirun_py.util.null_check import is_null_value

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

VALUE = 'value'
SOURCE = 'source'
KEY = 'key'


class ObjectDeleteKey(AbstractFunction):

    def __init__(self) -> None:
        super().__init__()
        self._signature = (
            FunctionSignature('ObjectDeleteKey')
            .set_namespace(Namespaces.SYSTEM_OBJECT)
            .set_parameters(
                dict([
                    Parameter.of_entry(SOURCE, Schema.of_any(SOURCE)),
                    Parameter.of_entry(KEY, Schema.of_string(KEY)),
                ])
            )
            .set_events(
                dict([Event.output_event_map_entry({VALUE: Schema.of_any(VALUE)})])
            )
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(self, context: FunctionExecutionParameters) -> FunctionOutput:
        source = context.get_arguments().get(SOURCE) if context.get_arguments() else None
        key = context.get_arguments().get(KEY) if context.get_arguments() else None

        if is_null_value(source):
            return FunctionOutput([EventResult.output_of({VALUE: None})])

        source = duplicate(source)
        if isinstance(source, dict) and key in source:
            del source[key]

        return FunctionOutput([EventResult.output_of({VALUE: source})])
